python/ray/train/huggingface/huggingface_trainer.py

```diff
@@ -283,6 +281,58 @@ def __init__(
             resume_from_checkpoint=resume_from_checkpoint,
         )
 
+    @classmethod
+    def _create_trainer_init_config(
+        cls,
+        trainer_init_per_worker: Callable[
+            [TorchDataset, Optional[TorchDataset], Any],
+            transformers.trainer.Trainer,
+        ],
+        trainer_init_config: Optional[Dict[str, Any]],
+    ) -> Dict[str, Any]:
+        trainer_init_config = trainer_init_config.copy() if trainer_init_config else {}
+        if TRAINER_INIT_FN_KEY in trainer_init_config:
+            raise ValueError(
+                f"'{TRAINER_INIT_FN_KEY}' is a reserved key in `trainer_init_config`."
+            )
+        if trainer_init_per_worker:
+            trainer_init_config[TRAINER_INIT_FN_KEY] = trainer_init_per_worker
+        return trainer_init_config
+
+    @classmethod
+    def restore(
```

Comment:
it kinda bothers me a lot now that we have to create a custom restore() for every single Trainer we have.
nobody would understand how to contribute a new Trainer if we do this ...

Reply:
What about:
Allow user to re-specify everything in `__init__` except for the run config. We can inspect the arguments of `cls.__init__` to see what's allowed. Then just overwrite everything that's passed in. Now, the subclasses don't need to re-implement restore.

One downside is that validation will be harder, and the signature of restore will be something like `def restore(cls, **kwargs):` which is very opaque.

Reply:
Latest iteration: User subclass doesn't need to set any attributes themselves. They just need to define what attributes from their constructor can get re-specified. Then, they just pass these through as kwargs to the `BaseTrainer.restore`.

```python
class MyTrainerSubclass(BaseTrainer):
    def restore(cls, path, my_arg, ...):
        return super().restore(path, my_arg=my_arg, ...)
        # No more setting things afterwards.
```

Reply:
this is nice!